### DISCUSSION TITLE
[GUIDE] Database problems on updating? Here's a workaround

### DISCUSSION THREAD

**Discussion by Darthagnon** [MikeMaximus/gbm#227] [General]
## Problem
1. While v1.25 was in development, I was helping with icons and giving feedback/ideas around the GUI. I cloned the source code and built the EXE in Visual Studio.
2. The homebrew build worked, and I used it for a while, as I kinda liked seeing all datafields in one page, vs. tabs.
3. The next versions officially released, and I tried to update GBM.
4. However, when running the updated version, scary-looking database error warnings appeared, and it didn't work. So I stuck with the old homebrew version for a few months.

## Solution
28/04/2021, I decided to update GBM from v1.25beta homebrew to v1.27 official. There were still the error windows popping up, so here's how I fixed it:

1. From the old v1.25beta homebrew installation, I clicked ```Setup``` >> ```Game Manager``` >> ```Export```, which exported an XML with all the game metadata and paths I'd entered.
2. I went to ```C:\Users\Darthagnon\Documents\Game Backup Monitor``` and renamed the folder to ```Game Backup Monitorbak``` (this makes a safe backup of my game save backups, just in case)
3. I went to ```C:\Users\Darthagon\AppData\Local\gbm``` and moved all the ```s3db``` database files into a ```bak``` folder, so they wouldn't be detected
4. Just to be safe, I ran the portable version of GBM v1.27, and started it up as a blank, new setup.
5. Once I got to the Game Manager, I clicked on ```Import``` and imported the previously exported XML with the game metadata. This restored the games list and data, but not the save backups.
6. Then, from the main window, I clicked ```Tools``` >> ```Import Backup``` >> ```Folder``` and selected ```C:\Users\Darthagnon\Documents\Game Backup Monitorbak```, which imported all the save file backups into the new empty Game Backup Monitor folder.
7. Then, I went back to the bak folders and renamed the s3db files to ```gbm.s3db.launch.v125betacorrupt.bak``` and ```gbm.s3db.v125betacorrupt.bak```, and moved all the ```*.bak``` database backup files from previous versions back into the main folder (keeping them just in case?)

**Comment by Darthagnon** [comment]
Question: Do I need to keep a copy of an exported config XML around? Or, provided I don't use homebrew/in-development builds, should GBM autodetect all the games list, backup files and settings, when e.g. restoring on a new PC?

**Reply by MikeMaximus** [reply]
I don't think it's necessary, but it's never bad to have an extra backup of your data around.

Situations like what you encountered should only ever be an issue when building from source during development of a new version before database changes are finalized.  

On a new system in a normal situation,  all you'd have to do is point GBM at your existing backup folder during the start-up wizard and it'll just import everything automatically.

**Comment by MikeMaximus** [comment]
Yes this works if you get into a situation where your GBM database files are rendered unusable or can't be upgraded correctly.  But you'll still lose some data such as sessions, linked processes, launcher configurations etc.

Simplifying a little bit:

1.  If possible, export your existing configurations from the Game Manager to an XML file.
2.  Make sure GBM is closed,  then delete the gbm.s3db files in the backup folder and %LOCALAPPDATA%\gbm (or move them to another location).
3.  Start GBM and it will assume it's a fresh installation.  Point it at your existing backup folder during the Start-Up Wizard.  This will automatically import the backup files and the configurations from the backup metadata if they were made with v1.2.0 or later.
4.  If available, import your configurations from the XML file in the Game Manager.  This restores any configurations that may not have been imported automatically from the backup folder.

Thanks for posting your solution 👍

**Reply by Darthagnon** [reply]
Thank you so much for the simplification! There was a fair bit of redundancy in my guide covering the steps I took, as I wasn't sure how GBM would react to the backup folder having saves in it, but not having the XML imported yet.

I did notice I lost playtime, but that wasn't a particularly important stat for me,

**Comment by MikeMaximus** [comment]
I just wanted to make a note here that in-development branches will now state in the readme.txt of that branch if they currently backwards compatible with the prior version or not.  

That way you'll know if it's currently safe to mess around with a development build without the risk of running into an issue like this.
